client/logmon/logmon_test.go

```diff
@@ -0,0 +1,120 @@
+package logmon
+
+import (
+	"bytes"
+	"crypto/rand"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/nomad/client/lib/fifo"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLogmon_Start_rotate(t *testing.T) {
+	require := require.New(t)
+	dir, err := ioutil.TempDir("", "nomadtest")
+	require.NoError(err)
+	defer os.RemoveAll(dir)
+	stdoutLog := "stdout"
+	stdoutFifoPath := filepath.Join(dir, "stdout.fifo")
+	stderrLog := "stderr"
+	stderrFifoPath := filepath.Join(dir, "stderr.fifo")
+
+	cfg := &LogConfig{
+		LogDir:        dir,
+		StdoutLogFile: stdoutLog,
+		StdoutFifo:    stdoutFifoPath,
+		StderrLogFile: stderrLog,
+		StderrFifo:    stderrFifoPath,
+		MaxFiles:      2,
+		MaxFileSizeMB: 1,
+	}
+
+	lm := NewLogMon(testlog.HCLogger(t))
+	require.NoError(lm.Start(cfg))
+
+	stdout, err := fifo.Open(stdoutFifoPath)
+	require.NoError(err)
+
+	// Write enough bytes such that the log is rotated
+	bytes1MB := make([]byte, 1024*1024)
+	_, err = rand.Read(bytes1MB)
+	require.NoError(err)
+
+	io.Copy(stdout, bytes.NewBuffer(bytes1MB))
+
+	time.Sleep(200 * time.Millisecond)
```

Comment:
I think you underestimate how slow CI can be. :sweat_smile: 

Use https://godoc.org/github.com/hashicorp/nomad/testutil#WaitForResult or similar.

Probably assert the size of the file as well and/or that `stdout.2` doesn't exist? I forget if we attempt to rotate at newline boundaries, but if so that will make the size nondeterministic.

You could switch the rand.Read to `bytes.Repeat([]byte("line\n"), 21000)` to make the output file sizes fixed if so.